### ISSUE TITLE
Question - Even more compact compact blocks?

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#668]
We are proposing "compact blocks" that contains "short id" representations of all inputs, outputs and kernels.
We would look in the tx pool and examine inputs, outputs and kernels of txs in the pool to identify the contents of the block to allow us to "hydrate" a compact block back into a full block.
This reduces the network traffic as we do not need to propagate full tx data twice (one in the tx and once in the block).

Transaction pool cut-through introduces some competing and conflicting needs here - different nodes may have different views of the txs (potentially cut-through in different ways) and nodes may not have the individual inputs and outputs available in the tx pool.

Note - all kernels would be available as these are never subject to cut-through.

And this point (kernels never get cut-through) is making me wonder if we cannot go a step further with compact blocks and potentially simplify the problem with tx pool cut-through in the process.

My question is - can we reduce a compact block down to simply the short_id of each contained tx kernel (and omit the short_ids for the inputs and outputs entirely)?
We would look in the tx pool for txs that contained the specified tx kernels (based on short_ids).
The block would then be reconstructed using the _inputs_, _outputs_ and _kernels_ of those txs from the pool, identified by the _kernel_ short_ids.
If the block is valid after reconstruction it would seem to follow that we are good.
It seems really unlikely that we could have a block built from txs based on a set of tx kernels that was valid, yet (after cut-through either at the tx pool layer or cut-through in the block itself) was a _different_ block to the one originally propagated.
This is just intuition though - I'm not sure this works in practice.

@ignopeverell thoughts on this?

Simple example - 

Tx<sub>1</sub> with [In<sub>1</sub>], [Out<sub>2</sub>], [Kern<sub>1</sub>].
Tx<sub>2</sub> with [In<sub>2</sub>], [Out<sub>3</sub>], [Kern<sub>2</sub>].

These could be cut-through to produce (In<sub>2</sub> spends Out<sub>2</sub>) - 
Tx<sub>3</sub> with [In<sub>1</sub>], [Out<sub>3</sub>], [Kern<sub>1</sub>, Kern<sub>2</sub>].

A block containing Tx<sub>3</sub> would be _identical_ to a block containing [Tx<sub>1</sub>, Tx<sub>2</sub>] after cut-through.

And that block in compact block form would contain short_ids for [Kern<sub>1</sub>, Kern<sub>2</sub>].

If a node had Tx<sub>3</sub> in the tx pool then these short_ids would find this single tx.
Alternatively if a node has both Tx<sub>1</sub> and Tx<sub>2</sub> in the tx pool then we would find different txs but the end result (the inputs, outputs and kernels) would be _exactly the same_ after cut-through. Regardless of if cut-through occurred in the tx pool _or_ in the block itself.

Put another way - we identify txs by their constituent kernels, by short_id.

What do we lose if we only include kernel short_ids in a compact block? And rely on the tx pool lookup to identify the inputs and outputs (based on txs containing these kernels)?
Does this introduce ambiguity or a greater chance of collisions?

**Comment by antiochp** [comment]
I think there are still cases where the tx pool of a node contains a tx that is "even more cut-through" than the sending node, where we cannot untangle the txs - in which case this does not totally solve the tx pool issue.
But this question is more around the compact block part of this - can we safely represent a compact block with kernel short_ids only?

**Comment by ignopeverell** [comment]
I think so. There are a couple concerns:

1. As you mention, pre-block cut-through may complicate things a bit.
2. This implicitly introduces a tx id. We've been avoiding this so far because people may latch on to this idea and use it as a way to identify their transactions, which is a bad idea once it's in a block.

However the benefits may still be worth it. Regarding the first point, we don't know what pre-block cut-through will look like so I wouldn't worry too much about it until we do. The 2nd point may be a little more concerning but we can just not advertize this too much. Kernels are part of the model regardless and we can't remove them anyway.

Bottom line, overall I'd be in favor of it.

**Comment by antiochp** [comment]
Awesome.
I think we can reinforce the idea these are Tx _Kernel_ Ids (and somewhat decoupled from Txs themselves).
Particularly if we end up tweaking Txs to allow multiple kernels in there.

Agreed on the other points about cut-through complexity and unknowns.

A 1,000 tx block in compact form is going to be pretty concise if we only need to include ids for the 1,000 kernels (and we can skip the x,000s of inputs and outputs).

**Comment by antiochp** [comment]
There's actually no way of avoiding the fact that these kernel ids _could_ be used to identify and track transaction right?
